### ISSUE TITLE
Removing a provider alias shouldn't fail if all its resources are removed at the same time

### ISSUE THREAD

**Issue by defreng** [hashicorp/terraform#30179]
### Use-cases
Removing a provider configuration in the same `terraform apply` as removing its resources results in a failure:
```
│ Error: Provider configuration not present
│ 
│ To work with null_resource.test (orphan) its original provider configuration at
│ provider["registry.terraform.io/hashicorp/null"].tp is required, but it has been removed.
│ This occurs when a provider configuration is removed while objects created by that provider
│ still exist in the state. Re-add the provider configuration to destroy null_resource.test
│ (orphan), after which you can remove the provider configuration again.
```

#### Why useful
Especially when working with the Vault provider, it's often necessary to create additional provider configurations to work with child namespaces.
Removing the resources in the child namespace and the provider at the same time unfortunately isn't possible with current terraform due to this issue

#### Minimal Example
A minimal example terraform code could look like:
```
provider "null" {
    alias = "tp"
}

resource "null_resource" "test" {
    provider = null.tp
}
```

1. Run `terraform apply` once to create the resources
2. Remove both the provider and the resource at the same time from the code
3. Run `terraform apply` again. It will fail with the mentioned error

### Proposal
I was wondering if it wouldn't be possible to keep the provider configuration stored in terraform state as well? So it would still be accessible in the case of removal.

**Comment by apparentlymart** [comment]
Hi @defreng! Thanks for sharing this use-case.

Terraform has historically not stored provider configurations in the state because they almost always contain credentials. Aside from the obvious concern that credentials can be sensitive, in many systems (including Vault) credentials are also _time-limited_, and so storing them in the state wouldn't necessarily help because the saved values are unlikely to still be valid by the next operation.

I do agree that it's unfortunate that Terraform requires the provider configuration to still be present during a destroy, but so far we've not found any suitable design to avoid that problem without introducing new problems. As you've seen, in current versions of Terraform it will always be necessary for a provider configuration to "outlive" all of the resources it manages, which does unfortunately sometimes call for certain kinds of changes requiring a multi-step process.

**Comment by crw** [comment]
Hi @defreng, I am going to close this ticket given this seems the issue as reported seems to be the intended or at least unavoidable behavior, and @apparentlymart offered some information which hopefully helps resolve or work-around the issue for you. Let me know if you'd like to dive into this in more detail and we can re-open this ticket.

**Comment by defreng** [comment]
Thanks @apparentlymart for the detailed explanation!

> Terraform has historically not stored provider configurations in the state because they almost always contain credentials. Aside from the obvious concern that credentials can be sensitive, in many systems (including Vault) credentials are also time-limited, and so storing them in the state wouldn't necessarily help because the saved values are unlikely to still be valid by the next operation.

I agree that this is a good reason not to store the entire provider configuration in the terraform state! Actually the whole problem you're describing wasn't obvious to me.

However (dream mode activated 🌌), when thinking a little longer about this it occured to my that provider configuration probably can be split very well into sensitive (credentials) and non-sensitive (aws region, vault namespace, ...) parts... Just from the way I experienced terraform usage so far, it typically resulted in the following pattern:
* non-sensitive configuration is typically provided explicitly in the provider configuration block
* sensitive configuration is typically provided via environment variables (either by native provider support like `VAULT_TOKEN`, or indirectly via terraform variables and `TF_VAR_*`)

Now, when removing the provider and resources using it at the same time, obviously the part stored in the configuration block is gone - but this part could've safely been stored in the terraform state as well. While typically (at least how we use it), all the required environment variables containing credentials are still present.

Long story short:
Maybe there would be a way to support the usecase described here by just storing the non-sensitive part of the provider configuration in the terraform state, and expecting the user to still supply (dynamic) credentials via env variables or cli arguments.

**Comment by apparentlymart** [comment]
I think what you are describing here is what I've typically described as the following two categories of settings:

1. Settings about _where things are_, like regions and endpoint URLs, which should typically remain constant between runs (unless intentionally changed).
2. Settings about _who or what is running Terraform_, like credentials, which are naturally scoped only to a particular run and so must be re-supplied every time.

We do typically recommend considering this distinction and only placing the first category of settings in your provider configuration, and have providers pick the second category up from platform-conventional "ambient" locations such as environment variables, credentials files, and so on.

The challenge of course is how to get from that being an optional recommendation to a hard requirement without requiring a tightly-coordinated change across Terraform Core and every provider all at once, and thereby forcing everyone to upgrade to the latest version of each provider in order to upgrade Terraform CLI. That's not viable, and so it's not been clear so far how to move into a world where that distinction is formalized in the system and therefore Terraform could safely rely on it to store only the first-category settings in the state/configuration, with the second-category settings always being provided from outside and expected to change on each run.

Of course, what I've written above is really just a more detailed version of the problem statement I shared before! In other words, the above is an example of what I meant when I said "without introducing new problems".

I would like to find suitable solutions to these challenges someday, but I don't expect we'll do so in the near future.
